gateway/join.go

```diff
@@ -0,0 +1,249 @@
+#include "gateway.h"
+#include <stdlib.h>
+
+*/
+import "C"
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto/cryptoservices"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/types"
+)
+
+type JoinRequest struct {
+	joinEUI  []byte
+	devEUI   []byte
+	devNonce []byte
+	mic      []byte
+}
+
+const (
+	joinRx2WindowSec = 6         // rx2 delay for sending join accept message.
+	rx2Frequenecy    = 923300000 // Frequuency for rx2 window
+	rx2SF            = 12        // spreading factor for rx2 window
+	rx2Bandwidth     = 0x06      // 500k bandwidth
+)
+
+// network id for the device to identify the network.
+var netID = []byte{1, 2, 3}
+
+func (g *Gateway) handleJoin(ctx context.Context, payload []byte) error {
+	jr, device, err := parseJoinRequestPacket(payload, g.devices)
+	if err != nil {
+		return err
+	}
+
+	joinAccept, err := device.GenerateJoinAccept(ctx, jr)
+	if err != nil {
+		return err
+	}
+
+	txPkt := C.struct_lgw_pkt_tx_s{
+		freq_hz:    C.uint32_t(rx2Frequenecy),
+		tx_mode:    C.uint8_t(0), // immediate mode
+		rf_chain:   C.uint8_t(0),
+		rf_power:   C.int8_t(26),    // tx power in dbm
+		modulation: C.uint8_t(0x10), // LORA modulation
+		bandwidth:  C.uint8_t(rx2Bandwidth),
+		datarate:   C.uint32_t(rx2SF),
+		coderate:   C.uint8_t(0x01), // code rate 4/5
+		invert_pol: C.bool(true),    // Downlinks are always reverse polarity.
+		size:       C.uint16_t(len(joinAccept)),
+	}
+
+	var cPayload [256]C.uchar
+	for i, b := range joinAccept {
+		cPayload[i] = C.uchar(b)
+	}
+	txPkt.payload = cPayload
+
+	// send on rx2 window - opens 6 seconds after join request.
+	time.Sleep(joinRx2WindowSec * time.Second)
+
+	// lock so there is not two sends at the same time.
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	errCode := int(C.send(&txPkt))
+	if errCode != 0 {
+		return errors.New("failed to send join accept packet")
+	}
+
+	return nil
+}
+
+// payload of join request consists of
+// | MHDR | JOIN EUI | DEV EUI  |   DEV NONCE  | MIC   |
+// | 1 B  |   8 B    |    8 B   |     2 B      |  4 B  |
+func parseJoinRequestPacket(payload []byte, devices map[string]*Device) (JoinRequest, *Device, error) {
+	var joinRequest JoinRequest
+
+	// everything in the join request payload is little endian
+	joinRequest.joinEUI = payload[1:9]
+	joinRequest.devEUI = payload[9:17]
+	joinRequest.devNonce = payload[17:19]
+	joinRequest.mic = payload[19:23]
+
+	matched := &Device{}
+
+	// device.devEUI is in big endian - reverse to compare and find device.
+	devEUIBE := reverseByteArray(joinRequest.devEUI)
+
+	// match the dev eui to gateway device
+	for _, device := range devices {
+		if bytes.Equal(device.devEui, devEUIBE) {
+			matched = device
+		}
+	}
+
+	if matched.name == "" {
+		return JoinRequest{}, nil, errors.New("received join request from unknown device")
```

Comment:
Can we return the name given in the request?

Reply:
So there’s no info about the device that’s sending data to us? 

Is it possible we would hit this error if a user misconfigured their devices in Viam?

Reply:
The identifying info in the join request is the dev EUI. If the user put in the wrong EUI into the config they would hit this. They would also hit this if there is another lorawan device nearby that the user doesn't care about sending messages.

Reply:
so is it worth returning the request's EUI as a part of the error, to help the user identify if their device is misconfigured or if theres another device interacting with the network that they are not aware of? 

I do not know how common it is to have outside devices pinging a network on the same frequencies, but I figure if a user misconfigured a device then returning the EUI should at least speed up debugging.

if you don't think its worthwhile feel free to ignore

Reply:
It could be common if they are in an area with a lot of other lora sensors, I will add a debug log with the dev EUI.